File: code/c/array/simple-stats.c
/* stats-calc.c */

#include <stdio.h>
#include <math.h>
#include <string.h>

#define DATA_SIZE 10

// Calculate the sum of the values in the array
double sum(const double data[], int size)
{
    int i;
    double result = 0;
    
    for(i = 0; i < size; i++)
    {
        result += data[i];
    }
    
    return result;
}

// Calculate the mean of the values in the array
double mean(const double data[], int size)
{
    return sum(data, size) / size;
}

// Find the largest value in the array
double max(const double data[], int size)
{
    //todo: add logic here...
    return 0;
}

// Find the standard deviation of the values in the array
double variance(const double data[], int size)
{
    //todo: add logic here...
    return 0;
}

double read_double(const char *prompt)
{
    double result;
    
    printf("%s", prompt);
    while (scanf(" %lf", &result) != 1)
    {
        scanf("%*[^\n]");
        printf("Please enter a number.\n");
        printf("%s", prompt);
    }
    
    return result;
}

void populate_array(double data[], int size)
{
    int i;
    char prompt[17] = ""; // enough space for "Enter value 99: " + terminator
    char buffer[3] = ""; // enough space for "99" + terminator
    
    for(i = 0; i < size; i++)
    {
        // Ensure that the terminator is included in the copy
        // so that the later calls to strncat know where to
        // append their details. 
        strncpy(prompt, "Enter value ", 13); // 12 + terminator
        sprintf(buffer, "%d", (i + 1) % 100); // needs space for 3 (2 + terminator)
        strncat(prompt, buffer, 2); // takes 3 spaces, 2 + terminator
        strncat(prompt, ": ", 2); // takes 3 spaces, 2 + terminator
        
        data[i] = read_double(prompt);
    }
}

// Implements a statistics calculator. The program reads in values entered by the user
// and then calculates the sum, mean, variance, and max
int main()
{
    double data[DATA_SIZE];
    
    populate_array(data, DATA_SIZE);
    
    printf("\nCalculating statistics...\n\n");
    
    printf("Sum:        %4.2f\n", sum(data, DATA_SIZE));
    printf("Mean:       %4.2f\n", mean(data, DATA_SIZE));
    printf("Variance:   %4.2f\n", variance(data, DATA_SIZE));
    printf("Max:        %4.2f\n", max(data, DATA_SIZE));
    
    return 0;
}

File: code/pascal/array/SimpleStats.pas
program StasticsCalculator;
uses SysUtils;

const DATA_SIZE = 10;

// Calculate the Sum of the values in the array
function Sum(const data: array of Double): Double;
var
  i: Integer;
begin
    result := 0;
    
    for i := Low(data) to High(data) do
    begin
        result += data[i];
    end;
end;

// Calculate the Mean of the values in the array
function Mean(const data: array of Double): Double;
begin
    result := Sum(data) / Length(data);
end;

// Find the largest value in the array
function Max(const data: array of Double): Double;
begin
    //todo: add logic here...
    result := 0;
end;

// Find the standard deviation of the values in the array
function Variance(const data: array of Double): Double;
begin
    //todo: add logic here...
    result := 0;
end;

function ReadDouble(prompt: String): Double;
var
  line: String;
begin
    Write(prompt);
    ReadLn(line);
    while not TryStrToFloat(line, result) do
    begin
        WriteLn('Please enter a number.\n');
        Write(prompt);
        ReadLn(line);
    end;
end;

procedure PopulateArray(var data: array of Double);
var
  i: Integer;
begin
    for i := Low(data) to High(data) do
    begin
        data[i] := ReadDouble('Enter value ' + IntToStr(i) + ': ');
    end;
end;

// Implements a statistics calculator. The program reads in values entered by the user
// and then calculates the Sum, Mean, variance, and max
procedure Main();
var
  data: array [0..DATA_SIZE-1] of Double;
begin
    PopulateArray(data);
    
    WriteLn('Calculating statistics...');
    
    WriteLn('Sum:        ', Sum(data):4:2);
    WriteLn('Mean:       ', Mean(data):4:2);
    WriteLn('Variance:   ', variance(data):4:2);
    WriteLn('Max:        ', max(data):4:2);
end;

begin
    Main();
end.